### PR TITLE
CESM Updates

### PR DESCRIPTION
**Pull request by dabail10**
For detailed information about submitting Pull Requests (PRs) to the CICE-Consortium,
please refer to: <https://github.com/CICE-Consortium/About-Us/wiki/Resource-Index#information-for-developers>

## PR checklist
- [X] Short (1 sentence) summary of your PR: 
Add some CESM changes to the NUOPC/CMEPS cap for async i/o and mask checking.
- [X] Developer(s): 
dabail10 (D. Bailey)
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
CESM aux_cice suite is bfb.
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please provide any additional information or relevant details below:
Some of the initialization in the cap is rearranged for the purposes of asynchronous i/o. Also, we have added a check for the mask comparing the grid file to the "mesh" file.

### CODE REVIEW

**Comment by DeniseWorthen** [comment]
@dabail10 I see two issues. First, the call to advertise_fields is never reached for UFS---the call has been moved inside of an CESMCOUPLED ifdef block. The second issue is that I'm not sure moving the advertise_fields prior to the call to cice_init2 works. It is in cice_init2 that the ncat is set, which is needed for the initialization of the export fields which contain categories, right?

File: cicecore/drivers/nuopc/cmeps/ice_import_export.F90 [hunk 1]
```diff
@@ -435,6 +436,7 @@ subroutine ice_import( importState, rc )
     integer                          :: ilo, ihi, jlo, jhi !beginning and end of physical domain
     type(block)                      :: this_block         ! block information for current block
     real (kind=dbl_kind),allocatable :: aflds(:,:,:,:)
+    real (kind=dbl_kind), dimension(nx_block,ny_block,max_blocks) :: work
```
Comment (DeniseWorthen) [inline cicecore/drivers/nuopc/cmeps/ice_import_export.F90:439]: This variable doesn't seem to be used. Can it be removed?

**Comment by dabail10** [comment]
Thanks for testing this! We know this works in CESM, but that is why I asked you to look at it for the UFS system. Let me look at the changes and I will hopefully have something again soon.

**Comment by DeniseWorthen** [comment]
Just about to comment. It does look like cice_init1 will initialize the parameters I was concerned about (ncat, nfreq, nfsd) through the call to input_data. So I think the only change required will be to add the call to advertise_fields outside of the cesm ifdef block in InitializeAdvertise. I'm testing that now.

**Comment by DeniseWorthen** [comment]
@dabail10 I've tested these two changes and everything passes for UFS.

```
diff --git a/cicecore/drivers/nuopc/cmeps/ice_comp_nuopc.F90 b/cicecore/drivers/nuopc/cmeps/ice_comp_nuopc.F90
index 4d82102..41ca095 100644
--- a/cicecore/drivers/nuopc/cmeps/ice_comp_nuopc.F90
+++ b/cicecore/drivers/nuopc/cmeps/ice_comp_nuopc.F90
@@ -641,6 +641,9 @@ subroutine InitializeAdvertise(gcomp, importState, exportState, clock, rc)
     call cice_init1
     call t_stopf ('cice_init1')

+    ! Advertise fields
+    call ice_advertise_fields(gcomp, importState, exportState, flds_scalar_name, rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
 #endif

     !----------------------------------------------------------------------------
diff --git a/cicecore/drivers/nuopc/cmeps/ice_import_export.F90 b/cicecore/drivers/nuopc/cmeps/ice_import_export.F90
index 1b9f340..e4db010 100644
--- a/cicecore/drivers/nuopc/cmeps/ice_import_export.F90
+++ b/cicecore/drivers/nuopc/cmeps/ice_import_export.F90
@@ -436,7 +436,6 @@ subroutine ice_import( importState, rc )
     integer                          :: ilo, ihi, jlo, jhi !beginning and end of physical domain
     type(block)                      :: this_block         ! block information for current block
     real (kind=dbl_kind),allocatable :: aflds(:,:,:,:)
-    real (kind=dbl_kind), dimension(nx_block,ny_block,max_blocks) :: work
     real (kind=dbl_kind)             :: workx, worky
     real (kind=dbl_kind)             :: MIN_RAIN_TEMP, MAX_SNOW_TEMP
     real (kind=dbl_kind)             :: Tffresh
```

**Comment by dabail10** [comment]
Great! I will add these changes.

**Comment by dabail10** [comment]
Ok. @DeniseWorthen can you retest this to make sure I did the right thing?

**Review by DeniseWorthen — APPROVED** [review]

**Comment by DeniseWorthen** [comment]
Changes are good. Thanks.

**Comment by eclare108213** [comment]
Is this PR ready to be merged?

**Comment by apcraig** [comment]
@dabail10, just want to confirm this is ready to merge from your perspective.

**Comment by dabail10** [comment]
I believe this is ready to go.

**Pull request merged**